timemachine/md/hrex.py

```diff
@@ -152,6 +152,31 @@ def symmetrize(x):
     return 1 / (1 - mu_2)
 
 
+def get_samples_by_iter_by_replica(
+    samples_by_state_by_iter: List[List[Samples]], replica_idx_by_state_by_iter: List[List[ReplicaIdx]]
+) -> List[List[Samples]]:
+    """Permute and reshape samples returned by :py:func:`run_hrex` having shape
+
+        (hrex_iter, state_idx) -> samples
+
+    to a more convenient form for analyzing the trajectories of individual replicas
+
+        (replica_idx, hrex_iter) -> samples
+    """
+
+    def inverse_permutation(p):
+        return [i for i, _ in sorted(enumerate(p), key=lambda t: t[1])]
+
+    samples_by_replica_by_iter = [
+        [samples_by_state[state_idx] for state_idx in inverse_permutation(replica_idx_by_state)]
+        for samples_by_state, replica_idx_by_state in zip(samples_by_state_by_iter, replica_idx_by_state_by_iter)
+    ]
+
+    samples_by_iter_by_replica = [list(xs) for xs in zip(*samples_by_replica_by_iter)]  # transpose
```

Comment:
nit: verify that the length of all lists in `samples_by_replica_by_iter` are the same length to avoid zip using the shortest length list and dropping data. Perhaps also for the samples list and the replicas list as they also get zipped

Reply:
Good call, added assertions in https://github.com/proteneer/timemachine/pull/1217/commits/0be0d574f8233c935539c1fd6530c4c82a3caa4e and https://github.com/proteneer/timemachine/pull/1217/commits/53747236adc86a7382461bc0a91a717866741bf7